libs/sql-parser/src/main/java/io/crate/sql/Identifiers.java

```diff
@@ -132,13 +132,25 @@ private static boolean charIsOutsideSafeRange(int i, char c) {
     }
 
     private static boolean quotesRequired(String identifier) {
-        return isKeyWord(identifier) || !IDENTIFIER.matcher(identifier).matches();
+        return isKeyWord(identifier) ||
+               // schema names like s, S, "s" are resolved to s (lower case), while "S" (quoted, mixed case) is resolved to S (upper case).
+               // Therefore, mixed case schema names should be quoted whenever necessary. (i.e. printed as a string and to be re-parsed)
+               !IDENTIFIER.matcher(identifier).matches() || isMixedCase(identifier);
```

Comment:
Isn't this now partially redundant? Maybe it would be possible to get rid of the identifier pattern match. It's fairly expensive compared to the mixed case check

Reply:
Thanks for reviewing. If it's ok, I will look at this separately. I think it is somewhat related to `quote()`, `escape()` and `SqlFormatter`, `ExpressionFormatter`.

Reply:
Yes fine to follow up on this, it's probably micro optimization anyways.

Easy variant would probably loop through the char and check for either mixed ascii or anything outside